Kernel/Problem.hpp

```diff
@@ -67,23 +70,70 @@ class Problem {
   bool hadIncompleteTransformation() const { return _hadIncompleteTransformation; }
   void reportIncompleteTransformation() { _hadIncompleteTransformation = true; }
 
-  typedef DHMap<unsigned,bool> TrivialPredicateMap;
-  void addTrivialPredicate(unsigned pred, bool assignment);
+  enum class IntereferenceKind : unsigned char {
+    FUN_DEF = 0,
+    PRED_DEF = 1,
+    GLOB_FLIP = 2,  // backs up random_polarities
+    COND_FLIP = 3,  // backs up partially  predicates (Plaisted-Greenbaum-style p<=>q reduced to, e.g. p=>q) and eliminated blocked clauses
+  };
+
+  /* Intereference (satisfiability-preserving-only transformations) recording */
+  struct Intereference {
+    IntereferenceKind _kind;
+    Intereference(IntereferenceKind kind) : _kind(kind) {}
+
+    virtual void outputDefinition(std::ostream&) = 0;
+  };
+
+  struct FunDef : public Intereference {
+    Term* _head;
+    Term* _body;
+    FunDef(Term* head, Term* body) : Intereference(IntereferenceKind::FUN_DEF), _head(head), _body(body) {}
+
+    void outputDefinition(std::ostream&) override;
+  };
+
+  struct PredDef : public Intereference {
+    Literal* _head;
+    Formula* _body;
+    PredDef(Literal* head, Formula* body) : Intereference(IntereferenceKind::PRED_DEF), _head(head), _body(body) {}
+
+    void outputDefinition(std::ostream&) override;
+  };
+
+  struct GlobalFlip : public Intereference {
+    unsigned _pred;
+
+    GlobalFlip(unsigned pred) : Intereference(IntereferenceKind::GLOB_FLIP), _pred(pred) {}
+
+    void outputDefinition(std::ostream&) override;
+  };
+
+  struct CondFlip : public Intereference {
+    Formula* _cond;
+    bool _neg;
+    Literal* _val;
+    bool _fixedPoint;
+
+    CondFlip(Formula* cond, bool neg, Literal* val, bool fixedPoint = false)
+      : Intereference(IntereferenceKind::COND_FLIP), _cond(cond), _neg(neg), _val(val), _fixedPoint(fixedPoint) {}
+
+    void outputDefinition(std::ostream&) override;
+  };
+
   /**
-   * Return map of trivial predicates into their assignments.
-   *
-   * Trivial predicates are the predicates whose all occurrences
-   * can be assigned either true or false.
-   */
-  const TrivialPredicateMap& trivialPredicates() const { return _trivialPredicates; }
+   * Pushing here during preprocessing,
+   * replaying backwards to get a model of the original signature.
+  */
+  Stack<Intereference*> intereferences;
```

Comment:
Could this be `unique_ptr`'d or otherwise managed?

Reply:
`std::vector` may help, I'm not sure how correct `Stack` is yet for move semantics, weird initialisers and so on. Also, death to homebrew containers.